lib/kube/proxy/streamproto/proto.go

```diff
@@ -0,0 +1,280 @@
+			return nil, trace.Wrap(err)
+		}
+
+		if ty != websocket.TextMessage {
+			return nil, trace.Errorf("Expected websocket control message, got %v", ty)
+		}
+
+		var msg metaMessage
+		if err := utils.FastUnmarshal(data, &msg); err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		if msg.ClientHandshake == nil {
+			return nil, trace.Errorf("Expected websocket client handshake")
+		}
+
+		s.Mode = msg.ClientHandshake.Mode
+	}
+
+	go s.readTask()
+	return s, nil
+}
+
+func (s *SessionStream) readTask() {
+	for {
+		terminated := false
+		defer s.closedC.Do(func() { close(s.done) })
+		ty, data, err := s.conn.ReadMessage()
+		if err != nil {
+			return
+		}
+
+		if ty == websocket.BinaryMessage {
+			s.in <- data
+		}
+
+		if ty == websocket.TextMessage {
+			var msg metaMessage
+			if err := utils.FastUnmarshal(data, &msg); err != nil {
+				return
+			}
+
+			if msg.Resize != nil {
+				s.resizeQueue <- msg.Resize
+			}
+
+			if msg.ForceTerminate && !terminated {
+				terminated = true
+				close(s.forceTerminate)
+			}
+		}
+
+		if ty == websocket.CloseMessage {
+			s.conn.Close()
+			s.closed = true
+			return
+		}
+	}
+}
+
+func (s *SessionStream) Read(p []byte) (int, error) {
+	if len(s.currentIn) == 0 {
+		select {
+		case s.currentIn = <-s.in:
+		case <-s.done:
+			return 0, io.EOF
+		}
+	}
+
+	n := copy(p, s.currentIn)
+	s.currentIn = s.currentIn[n:]
+	return n, nil
+}
+
+func (s *SessionStream) Write(data []byte) (int, error) {
+	s.writeSync.Lock()
+	defer s.writeSync.Unlock()
+
+	err := s.conn.WriteMessage(websocket.BinaryMessage, data)
+
+	if err != nil {
+		return len(data), s.conn.WriteMessage(websocket.BinaryMessage, data)
+	}
+
+	return 0, trace.Wrap(err)
+}
+
+// Resize sends a resize request to the other party.
+func (s *SessionStream) Resize(size *remotecommand.TerminalSize) error {
+	msg := metaMessage{Resize: size}
+	json, err := utils.FastMarshal(msg)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	s.writeSync.Lock()
+	defer s.writeSync.Unlock()
+	return trace.Wrap(s.conn.WriteMessage(websocket.TextMessage, json))
+}
+
+// ResizeQueue returns a channel that will receive resize requests.
+func (s *SessionStream) ResizeQueue() <-chan *remotecommand.TerminalSize {
+	return s.resizeQueue
+}
+
+// ForceTerminateQueue returns the channel used for force termination requests.
+func (s *SessionStream) ForceTerminateQueue() <-chan struct{} {
+	return s.forceTerminate
+}
+
+// ForceTerminate sends a force termination request to the other end.
+func (s *SessionStream) ForceTerminate() error {
+	msg := metaMessage{ForceTerminate: true}
+	json, err := utils.FastMarshal(msg)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	s.writeSync.Lock()
+	defer s.writeSync.Unlock()
+
+	return trace.Wrap(s.conn.WriteMessage(websocket.TextMessage, json))
+}
+
+func (s *SessionStream) Done() <-chan struct{} {
+	return s.done
+}
+
+// Close closes the stream.
+func (s *SessionStream) Close() error {
+	if !s.closed {
+		s.closed = true
```

Comment:
This looks like a data race to me. Have you built this with the race detector on to verify?

`s.closed` appears to be read from the `readTask` goroutine. You can't write to it from this goroutine without proper synchronization, right?

Reply:
I did run it with a race detector. But this does actually seem to be a data race neither it or me caught. I've switched to using an atomic and changing the strategy slightly to resolve it.